garage/tf/samplers/off_policy_vectorized_sampler.py

```diff
@@ -58,7 +58,7 @@ def shutdown_worker(self):
         self.vec_env.close()
 
     @overrides
-    def obtain_samples(self, itr):
+    def obtain_samples(self, itr, batch_size=None):
```

Comment:
hmm is there an algorithm which wants this option?

Reply:
No there's not. `obtain_samples()` is only called by runner now and the batch_size is piped from `Runner.train()` to `Sampler.obtain_sample()`. I think it's better to make it an argument of `Runner.train()` rather than to fix it in sampler or runner constrctor, so that I can call `Runner.train()` multiple times with different batch_size.

Reply:
Okay I see what you're saying. One thing to keep in mind is that `batch_size` is an important hyperparameter for algorithms/experiments, so eventually it will be an important parameter of the experiment configuration. You need to make sure it can be set once for an experiment and reliably followed by the rest of the code (if that's not the case right now, it can be a fix for a future PR)

i guess what i'm really asking is:

_"what is the interface for `obtain_samples`?"_

it seems right now you have settled on

_`obtain_samples` returns `n_envs * max_path_length` samples unless a `batch_size` parameter is provided, in which case it provides `batch_size` samples_

This seems mostly fine (not going to block the PR on this), but **here's why I think it might be confusing**:
Using `batch_size = n_envs * max_path_length` will not necessarily return `n_envs` trajectories, because some trajectories will end early. Your sampler will keep going until it fills `n_samples >= batch_size`, which in general will be more than `n_envs` trajectories. 

If I encountered your API in the wild and called it with no `batch_size` parameter, I think I would expect to get back `n_envs` trajectories, with those trajectories having however many samples they have (some end early)--not some unpredictable number of trajectories `n_envs * max_path_length`. After all, if I wanted that, I could have just called it myself.

I think the tension here between two definitions of how many samples you want: do you want `batch_size` time steps or do you want some number of episodes `n_episodes`? Some algorithms want a fixed number of episode trajectories, and others want a fixed batch size. There's no reason, by the way, the desired number of  trajectory episodes must be `n_episodes == n_envs`. That's just a limitation of our current code.

So perhaps what really makes sense here is two options: you can ask `obtain_samples()` for either some number of episodes `n_episodes` OR some number of samples `n_samples` -- but you must provide me with one and exactly one of those numbers.

This is perhaps fodder for another PR, but just be aware it's a real issue. To avoid mangling arguments to determine behavior, you can always split the API: those wishing to be trajectory-oriented could call `def obtain_episodes(self, itr, n_episodes)` and those wishing to be sample-oriented could call `def obtain_samples(self, itr, n_samples)`

Reply:
I see your points on `batch_size`. Indeed, this parameter is not well defined and our current implementation doesn't have a clear logic with respect to it. The default value of `batch_size` is in fact copied from original `OffPolicyVectorizedSampler`. It's not configurable (but should be IMO)
https://github.com/rlworkgroup/garage/blob/9ba429f3f5e2c4593170aef61f38bf94b8d1fae1/garage/tf/samplers/off_policy_vectorized_sampler.py#L73

Let's make it future PR to offer the option between `obtain_episodes()` and `obtain_samples()`.

Reply:
This definition is like a quick fix for using VectorizedSampler, I have some concerns on sampling fixed samples. Can we discuss it on stand up?

Reply:
Sure thing. Let's talk tomorrow.